### PR TITLE
Code review request

### PR DESCRIPTION
**Pull request by GregMcLindon**
check formatting, naming and logic @Alan-Liang-1997 

Hi @david-micallef can you look at my miles_converter app. During the last part of the prac where we had to customise the Kivy widgets I can't get the text input to display in black... still accepts input so must be white font but can't get it to appear?

### CODE REVIEW

File: prac_04/quick_pick_solution_from_git.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+"""
+CP1404/CP5632 Practical - Suggested Solution
+Quick pick program
+"""
+
+import random
+
+NUMBERS_PER_LINE = 6
```
Comment (lindsaymarkward) [inline prac_04/quick_pick_solution_from_git.py]: Good constants 👍

File: prac_04/quick_pick_solution_from_git.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+def main():
+    """Quick picks program - choose sets of random numbers."""
+    quick_picks = int(input("How many quick picks? "))
+    while quick_picks < 0:
+        print("That makes no sense!")
+        quick_picks = int(input("How many quick picks? "))
+
+    for i in range(quick_picks):
+        for j in range(NUMBERS_PER_LINE):
+            number = random.randint(MINIMUM, MAXIMUM)
```
Comment (lindsaymarkward) [inline prac_04/quick_pick_solution_from_git.py]: Where do you ensure that the numbers are unique? (You need to do this.)

File: prac_04/quick_pick_solution_from_git.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            print("{:2}".format(number), end=' ')
```
Comment (lindsaymarkward) [inline prac_04/quick_pick_solution_from_git.py]: Where do you format your output so numbers line up?

File: prac_05/color_names.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+"""CP1404 Practical week 5: look up color codes"""
+
+color_names = {"AliceBlue": "#f0f8ff", "AntiqueWhite1": "#ffefdb", "azure4": "#838b8b", "beige": "#f5f5dc",
+               "bisque1": "#ffe4c4", "bisque2": "#eed5b7", "bisque3": "#cdb79e", "bisque4": "#8b7d6b",
+               "black": "#000000", "BlanchedAlmond": "#ffebcd", "cyan1": "#00ffff", "cyan2": "#00eeee",
+               "cyan3": "#00cdcd", "cyan4": "#008b8b", "DarkGoldenrod": "#b8860b", "DarkGoldenrod1": "#ffb90f",
+               "DarkGoldenrod2": "#eead0e"}
+color_name = input("Enter color name: ")
+while color_name!="":
+    print(color_names[color_name])
```
Comment (lindsaymarkward) [inline prac_05/color_names.py]: You need to handle this crashing.
`color_names` is just the plural of `color_name`, so it sounds like a list of names, but it's a dictionary.

File: prac_05/word_count.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+"""CP1404 Practicals week5 - print the frequency of given words in a given sentance"""
+
+text_dict = {}
+text_input = input("Text: ")
+max_word_length = 0
+for word in text_input.split():
+    if len(word) > max_word_length:
+        max_word_length = len(word)
+    try:
```
Comment (lindsaymarkward) [inline prac_05/word_count.py:9]: Good 👍

**Review by lindsaymarkward — COMMENTED** [review]
@GregMcLindon Use the imperative voice for commit messages: "add" not "adds". 
I accidentally made my comments part of a "review". I haven't used that before, so I'm seeing how that works now :)

**Comment by lindsaymarkward** [comment]
I don't have a solution for the white text in the TextInput. Let me know if you find one.

File: prac_04/quick_pick_solution_from_git.py (hunk 2 above)
Comment (GregMcLindon) [reply prac_04/quick_pick_solution_from_git.py]: I'm not sure where that quick_pick_solution_from_git.py came from.... 

Refer this code this for my submission:
https://github.com/GregMcLindon/cp1404_practicals/blob/feedback/prac_04/quick_pick.py

Unique handled in generate_numbers function. 

``if not random_number in game_numbers: 
                 game_numbers.append(random_number)

File: prac_04/quick_pick_solution_from_git.py (hunk 3 above)
Comment (GregMcLindon) [reply prac_04/quick_pick_solution_from_git.py]: THink I went off book and used:

```
for game_numbers in numbers_selected: 
            print(*game_numbers) 
```

(https://github.com/GregMcLindon/cp1404_practicals/blob/feedback/prac_04/quick_pick.py)

File: prac_05/color_names.py (hunk 4 above)
Comment (GregMcLindon) [reply prac_05/color_names.py]: Ok will do (don't have PyCharm on work PC).... I thought we were meant to only do exactly what the prac asked us to do. I don't recall validation being a requirement.

File: prac_04/quick_pick_solution_from_git.py (hunk 3 above)
Comment (lindsaymarkward) [reply prac_04/quick_pick_solution_from_git.py]: This doesn't line them up neatly as in the sample (requirement).

File: prac_05/color_names.py (hunk 4 above)
Comment (lindsaymarkward) [reply prac_05/color_names.py]: The requirement isn't explicit, but the prac says "Based on the state name example program above". The state name example doesn't crash, and neither should yours.

File: prac_07/programming_language.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+class ProgrammingLanguage:
+    """CP1403 programming language class"""
+
+    def __init__(self, name: str = "", typing: str = "", reflection: bool = False, year: int = "0000"):
```
Comment (lindsaymarkward) [inline prac_07/programming_language.py]: PyCharm should be warning you about extra spaces in this line. Reformat the code to see what it changes.

File: prac_07/programming_language.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        """initialise a language type
+
+        Reflection: boolean, False = 'No', True = 'Yes'
```
Comment (lindsaymarkward) [inline prac_07/programming_language.py]: This (True = Yes) seems redundant

File: prac_07/programming_language.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        """
+        self.name = name
+        self.typing = typing
+        self.reflection = reflection
+        self.year = year
+
+    def __is_dynamic__(self):
+        """returns True if programming language is dynamic (as opposed to static)"""
+        return "{}".format(True if self.typing == "Dynamic" else False)
```
Comment (lindsaymarkward) [inline prac_07/programming_language.py]: Doesn't this return a str? It should return a boolean. Drop the string formatting.

File: prac_04/quick_pick_solution_from_git.py (hunk 3 above)
Comment (GregMcLindon) [reply prac_04/quick_pick_solution_from_git.py]: Sorry mislead you. The function is formatting the numbers to two characters when adding them to the list:
random_number = str("{:2}".format(random.randint(1,45)))
            if not random_number in game_numbers:
                game_numbers.append(random_number)

I've just reviewed the solution given in the practicals solution branch and it uses a similiar method of formatting so will leave as is change it.

File: prac_05/color_names.py (hunk 4 above)
Comment (GregMcLindon) [reply prac_05/color_names.py]: variable names refactored and exception handler added.
while color_search != "":
    try:
        print(color_dict[color_search])
    except:
        print("Enter valid name from following list:")
        for color in sorted(color_dict):
            print("{:5}{}".format("",color))
    color_search = input("Enter color name: ")

File: prac_07/programming_language.py (hunk 6 above)
Comment (GregMcLindon) [reply prac_07/programming_language.py]: Fixed:
def __init__(self, name: str="", typing: str="", reflection: bool=False, year: int="0000"):

File: prac_07/programming_language.py (hunk 7 above)
Comment (GregMcLindon) [reply prac_07/programming_language.py]: removed

File: prac_07/programming_language.py (hunk 8 above)
Comment (GregMcLindon) [reply prac_07/programming_language.py]: changed to:
        return self.typing == "Dynamic"

File: prac_04/quick_pick_solution_from_git.py (hunk 3 above)
Comment (lindsaymarkward) [reply prac_04/quick_pick_solution_from_git.py]: Oh, I missed that. You should definitely be adding numbers, not strings. We want numbers in the list. You should store data as its "correct" format. If you want to display it differently (e.g. adding $ and , . for currency), then you format the actual value.

File: prac_07/programming_language.py (hunk 7 above)
Comment (lindsaymarkward) [reply prac_07/programming_language.py]: Make sure you use correct data. "0000" is not an int.

Comment (GregMcLindon) [reply prac_07/programming_language.py]: fixed

File: prac_04/quick_pick_solution_from_git.py (hunk 3 above)
Comment (GregMcLindon) [reply prac_04/quick_pick_solution_from_git.py]: fixed (game numbers stored as integers and print function rebuilt using join string formatting:
numbers_selected = generate_numbers(picks_selected)
                for game_numbers in numbers_selected:
                    print(" ".join("{:2}".format(number) for number in game_numbers))

**Pull request merged**